### DISCUSSION TITLE
Testing Fore apps

### DISCUSSION THREAD

**Discussion by timathom** [Jinntec/Fore#248] [General]
I'm not a software developer, so my experience with different testing frameworks is limited, but I've been looking at the Cypress E2E testing in Fore, and it strikes me as something that could be really useful for users of Fore in general as a way to test their apps. Maybe it's something that could be packaged with or executed from Fore Glass?

**Comment by JoernT** [comment]
I agree that E2E testing of pages is extremely useful especially when it comes to more complex ones with complex loading and submissions. However Cypress will require that you got some nodejs installed on your machine (when not using cloud option) and we can't invoke that directly from the browser afaik.

However it might help to have a project structure generated to put your Fore project and have testing around?

E.g. kickstart it with `npm run create-project`

getting
- root-folder
-- cypress (test folder)
-- src
--- pages
--- resources
---- CSS
---- ... (fonts, scripts, ...)

we would put a local node-based http server so can run and test on localhost easily and run the E2E tests.

running tests would than simply be `npx cypress open` once you have installed nodejs and run `npm i`

When you work is done you would simply take everything from the 'src' folder and deploy it on your webserver. What do you think? Would that be helpful?`

**Reply by timathom** [reply]
Yes, what you describe sounds great to me as a project workflow. It would be very handy to have things set up like that.

**Comment by DrRataplan** [comment]
Hey Tim,

Good one. E2E testing is great for more complex applications. Cypress needs a host environment, though you can also run it in the cloud. So you will have to install it in your own project. This is something we could make a 'basis' for, like an example Fore app.

What are your thoughts Joern?

**Reply by DrRataplan** [reply]
Ah, you were just a bit quicker then me in w writing your comment hehe

**Comment by JoernT** [comment]
Thinking about it it would probably the simplest thing to create a seperate repo 'fore-project', setup the suggested structure in there and have a package.json in place to pull everything with a single `npm i` so we get a folder structure to put things in and have your tests alongside. This could be done quite quickly.

**Comment by JoernT** [comment]
@timathom @DrRataplan please have a look at https://github.com/Jinntec/fore-project - i tried to quickly throw something together. Let me know if if you can follow the readme and your thoughts.

**Reply by timathom** [reply]
@JoernT, yes, at first glance, this seems straightforward to me! I'll try to provide feedback once I have a chance to try for myself.
